test/integration_test/basic_dmthin_test.go

```diff
@@ -22,124 +22,259 @@ import (
+		}
 	}
 	return cluster
 }
 
-var testDmthinCases = []types.TestCase{
-	{
-		/* Install DMthin by specifying only metadata device */
-		TestName:        "BasicInstallDmthinWithMetadataDevice",
-		TestrailCaseIDs: []string{""},
-		TestSpec: func(t *testing.T) interface{} {
-			provider := cloud_provider.GetCloudProvider()
-			cluster := defaultDmthinSpec(t)
-			cloudSpec := &corev1.CloudStorageSpec{}
-			cloudSpec.DeviceSpecs = provider.GetDefaultDataDrives()
-			cloudSpec.SystemMdDeviceSpec = provider.GetDefaultMetadataDrive()
-			cluster.Spec.CloudStorage = cloudSpec
-			return cluster
-		},
-		TestFunc: BasicInstallDmthin,
-	},
-	{
-		/* Install DMthin by specifying metadata and kvdb devices */
-		TestName:        "BasicInstallDmthinWithMetadataAndKVDBDevice",
-		TestrailCaseIDs: []string{""},
-		TestSpec: func(t *testing.T) interface{} {
-			provider := cloud_provider.GetCloudProvider()
-			cluster := defaultDmthinSpec(t)
-			cloudSpec := &corev1.CloudStorageSpec{}
-			cloudSpec.DeviceSpecs = provider.GetDefaultDataDrives()
+func getTestName(opt OptionsArr) string {
+	testName := "BasicInstallDmthinWith"
+	if opt[PxStoreV2Present] {
+		testName += "PxStoreV2"
+	}
+	if opt[MetadataPresent] {
+		testName += "Metadata"
+	}
+	if opt[KvdbDevicePresent] {
+		testName += "KVDB"
+	}
+	if opt[JournalPresent] {
+		testName += "Journal"
+	} else if opt[AutoJournalPresent] {
+		testName += "AutoJournal"
+	}
+	return testName
+}
+
+func generateSingleTestCase(o OptionsArr) types.TestCase {
+	testCase := types.TestCase{}
+	testCase.TestName = getTestName(o)
+
+	testCase.TestSpec = func(t *testing.T) interface{} {
+		provider := cloud_provider.GetCloudProvider()
+		cluster := defaultDmthinSpec(t, o[PxStoreV2Present])
+		cloudSpec := &corev1.CloudStorageSpec{}
+		cloudSpec.DeviceSpecs = provider.GetDefaultDataDrives()
+		if o[MetadataPresent] {
 			cloudSpec.SystemMdDeviceSpec = provider.GetDefaultMetadataDrive()
+		}
+		if o[JournalPresent] {
+			cloudSpec.JournalDeviceSpec = provider.GetDefaultJournalDrive()
+		}
+		if o[AutoJournalPresent] {
+			autoStr := "auto"
+			cloudSpec.JournalDeviceSpec = &autoStr
+		}
+		if o[KvdbDevicePresent] {
 			cloudSpec.KvdbDeviceSpec = provider.GetDefaultKvdbDrive()
-			cluster.Spec.CloudStorage = cloudSpec
-			return cluster
-		},
-		TestFunc: BasicInstallDmthin,
-	},
+		}
+
+		cluster.Spec.CloudStorage = cloudSpec
+		return cluster
+	}
+	testCase.TestFunc = BasicInstallDmthin
+	return testCase
 }
 
-func TestDMthinCloudDrivesBasic(t *testing.T) {
-	for _, testCase := range testDmthinCases {
+func generateTestCases(opt OptionsArr, idx uint) []types.TestCase {
+	var retVal []types.TestCase
+	if idx == uint(len(opt)) {
+		singleTestCase := generateSingleTestCase(opt)
+		retVal = append(retVal, singleTestCase)
+		return retVal
+	}
+	without := generateTestCases(opt, idx+1)
+	retVal = append(retVal, without...)
+
+	opt[idx] = true
+	with := generateTestCases(opt, idx+1)
+	retVal = append(retVal, with...)
+	return retVal
```

Comment:
Thanks for explaining the logic here. It's a bit complicated so some comments to explain what you are trying to do here would be helpful.